File: python/samples/v2_x/create_account_level_filter_set.py
#!/usr/bin/python
#
# Copyright 2017 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This example creates an account-level filter set.

An account-level filter set can be used to retrieve data for a specific
Authorized Buyers account, whether that be a bidder or child seat account.
"""


import argparse
from datetime import date
from datetime import datetime
from datetime import timedelta
import os
import sys
import uuid

sys.path.insert(0, os.path.abspath('..'))

from googleapiclient.errors import HttpError
import samples_util


_DATE_FORMAT = '%Y%m%d'
_FILTER_SET_NAME_TEMPLATE = 'bidders/%s/accounts/%s/filterSets/%s'
_OWNER_NAME_TEMPLATE = 'bidders/%s/accounts/%s'
_TODAY = date.today()
_VALID_ENVIRONMENTS = ('WEB', 'APP')
_VALID_FORMATS = ('DISPLAY', 'VIDEO')
_VALID_PLATFORMS = ('DESKTOP', 'TABLET', 'MOBILE')
_VALID_TIME_SERIES_GRANULARITIES = ('HOURLY', 'DAILY')

DEFAULT_ACCOUNT_RESOURCE_ID = 'ENTER_ACCOUNT_RESOURCE_ID_HERE'
DEFAULT_BIDDER_RESOURCE_ID = 'ENTER_BIDDER_RESOURCE_ID_HERE'
DEFAULT_FILTER_SET_RESOURCE_ID = 'FilterSet_%d' % uuid.uuid4()
DEFAULT_END_DATE = _TODAY.strftime(_DATE_FORMAT)
DEFAULT_START_DATE = (_TODAY - timedelta(days=7)).strftime(
    _DATE_FORMAT)


def main(ad_exchange_buyer, owner_name, body, is_transient):
  try:
    # Construct and execute the request.
    filter_set = ad_exchange_buyer.bidders().accounts().filterSets().create(
        ownerName=owner_name, isTransient=is_transient, body=body).execute()
    print 'FilterSet created for bidder: "%s".' % (owner_name)
    print filter_set
  except HttpError as e:
    print e


if __name__ == '__main__':

  def time_series_granularity_type(s):
    if s not in _VALID_TIME_SERIES_GRANULARITIES:
      raise argparse.ArgumentTypeError('Invalid TimeSeriesGranularity '
                                       'specified: "%s".' % s)
    return s

  def environment_type(s):
    if s not in _VALID_ENVIRONMENTS:
      raise argparse.ArgumentTypeError('Invalid Environment specified: "%s".'
                                       % s)
    return s

  def format_type(s):
    if s not in _VALID_FORMATS:
      raise argparse.ArgumentTypeError('Invalid Format specified: "%s".' % s)
    return s

  def platform_type(s):
    if s not in _VALID_PLATFORMS:
      raise argparse.ArgumentTypeError('Invalid Platform specified: "%s".'
                                       % platform_type)
    return s

  def valid_date(s):
    try:
      return datetime.strptime(s, _DATE_FORMAT).date()
    except ValueError:
      raise argparse.ArgumentTypeError('Invalid date specified: "%s".' % s)

  parser = argparse.ArgumentParser(
      description=('Creates an account-level filter set with the specified '
                   'options.'))
  # Required fields.
  parser.add_argument(
      '-a', '--account_resource_id', default=DEFAULT_ACCOUNT_RESOURCE_ID,
      help=('The resource ID of the bidders.accounts resource for which the '
            'filter set is being created. This will be used to construct the '
            'ownerName used as a path parameter for filter set requests. For '
            'additional information on how to configure the ownerName path '
            'parameter, see: https://developers.google.com/authorized-buyers/'
            'apis/reference/rest/v2beta1/bidders.accounts.filterSets/'
            'create#body.PATH_PARAMETERS.owner_name'))
  parser.add_argument(
      '-b', '--bidder_resource_id', default=DEFAULT_BIDDER_RESOURCE_ID,
      help=('The resource ID of the bidders resource for which the filter set '
            'is being created. This will be used to construct the ownerName '
            'used as a path parameter for filter set requests. For additional '
            'information on how to configure the ownerName path parameter, '
            'see: https://developers.google.com/authorized-buyers/apis/'
            'reference/rest/v2beta1/bidders.accounts.filterSets/create'
            '#body.PATH_PARAMETERS.owner_name'))
  parser.add_argument(
      '-r', '--resource_id', default=DEFAULT_FILTER_SET_RESOURCE_ID,
      help=('The resource ID of the filter set. Note that this must be '
            'unique. This will be used to construct the filter set\'s name. '
            'For additional information on how to configure a filter set\'s '
            'name, see: https://developers.google.com/authorized-buyers/apis/'
            'reference/rest/v2beta1/bidders.accounts.filterSets'
            '#FilterSet.FIELDS.name'))
  parser.add_argument(
      '--end_date', default=DEFAULT_END_DATE, type=valid_date,
      help=('The end date for the filter set\'s absoluteDateRange field, which '
            'will be accepted in this example in YYYYMMDD format.'))
  parser.add_argument(
      '--start_date', default=DEFAULT_START_DATE, type=valid_date,
      help=('The start date for the filter set\'s time_range field, which '
            'will be accepted in this example in YYYYMMDD format.'))
  # Optional fields.
  parser.add_argument(
      '-c', '--creative_id', required=False,
      help=('The ID of the creative on which to filter.'))
  parser.add_argument(
      '-d', '--deal_id', required=False,
      help=('The ID of the deal on which to filter.'))
  parser.add_argument(
      '-e', '--environment', required=False,
      type=environment_type,
      help=('The environment on which to filter.'))
  parser.add_argument(
      '-f', '--format', required=False,
      type=format_type,
      help=('The format on which to filter.'))
  parser.add_argument(
      '-p', '--platforms', required=False, nargs='*', type=platform_type,
      help=('The platforms on which to filter. The filters represented by '
            'multiple platforms are ORed together. Note that you may specify '
            'more than one using a space as a delimiter.'))
  parser.add_argument(
      '-s', '--seller_network_ids', required=False, nargs='*', type=int,
      help=('The list of IDs for seller networks on which to filter. The '
            'filters represented by multiple seller network IDs are ORed '
            'together. Note that you may specify more than one using a space '
            'as a delimiter.'))
  parser.add_argument(
      '-t', '--time_series_granularity', required=False,
      type=time_series_granularity_type,
      help=('The granularity of time intervals if a time series breakdown is '
            'desired.'))
  parser.add_argument(
      '--is_transient', required=False, default=True, type=bool,
      help=('Whether the filter set is transient, or should be persisted '
            'indefinitely. In this example, this will default to True.'))

  args = parser.parse_args()

  # Build the time_range as an AbsoluteDateRange.
  time_range = {
      'startDate': {
          'year': args.start_date.year,
          'month': args.start_date.month,
          'day': args.start_date.day
      },
      'endDate': {
          'year': args.end_date.year,
          'month': args.end_date.month,
          'day': args.end_date.day
      }
  }

  # Create a body containing the required fields.
  BODY = {
      'name': _FILTER_SET_NAME_TEMPLATE % (
          args.bidder_resource_id, args.account_resource_id, args.resource_id),
      # Note: You may alternatively specify relativeDateRange or
      # realtimeTimeRange.
      'absoluteDateRange': time_range
  }

  # Add optional fields to body if specified.
  if args.creative_id:
    BODY['creativeId'] = args.creative_id
  if args.deal_id:
    BODY['dealId'] = args.deal_id
  if args.environment:
    BODY['environment'] = args.environment
  if args.format:
    BODY['format'] = args.format
  if args.platforms:
    BODY['platforms'] = args.platforms
  if args.seller_network_ids:
    BODY['sellerNetworkIds'] = args.seller_network_ids
  if args.time_series_granularity:
    BODY['timeSeriesGranularity'] = args.time_series_granularity

  try:
    service = samples_util.GetService(version='v2beta1')
  except IOError, ex:
    print 'Unable to create adexchangebuyer service - %s' % ex
    print 'Did you specify the key file in samples_util.py?'
    sys.exit(1)

  main(service, _OWNER_NAME_TEMPLATE % (args.bidder_resource_id,
                                        args.account_resource_id),
       BODY, args.is_transient)

File: java/src/main/java/com/google/api/services/samples/adexchangebuyer/cmdline/v2_x/CreateAccountLevelFilterSet.java
/*
 * Copyright (c) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.google.api.services.samples.adexchangebuyer.cmdline.v2_x;

import com.google.api.client.googleapis.services.json.AbstractGoogleJsonClient;
import com.google.api.services.adexchangebuyer2.v2beta1.AdExchangeBuyerII;
import com.google.api.services.adexchangebuyer2.v2beta1.model.AbsoluteDateRange;
import com.google.api.services.adexchangebuyer2.v2beta1.model.Date;
import com.google.api.services.adexchangebuyer2.v2beta1.model.FilterSet;
import com.google.api.services.samples.adexchangebuyer.cmdline.BaseSample;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

/**
 * This sample illustrates how to create an account-level filter set.
 *
 * An account-level filter set can be used to retrieve data for a specific Authorized Buyers
 * account, whether that be a bidder or child seat account.
 */
public class CreateAccountLevelFilterSet extends BaseSample {
  private DateTimeFormatter df = DateTimeFormat.forPattern("yyyyMMdd");

  @Override
  public ClientType getClientType() {
    return ClientType.ADEXCHANGEBUYERII;
  }

  @Override
  public String getName() {
    return "Create Account-level Filter Set";
  }

  @Override
  public String getDescription() {
    return "Creates a new Account-level Filter Set with the specified options.";
  }

  @Override
  public void execute(AbstractGoogleJsonClient client) throws IOException {
    AdExchangeBuyerII adXClient = (AdExchangeBuyerII) client;

    DateTime defaultEndTime = DateTime.now();
    DateTime defaultStartTime = defaultEndTime.minusDays(7);

    // Required fields.
    String bidderResourceId = getStringInput("bidderResourceId", "Enter the Bidder resource ID");
    String accountResourceId = getStringInput("accountResourceId", "Enter the Account resource ID");
    String resourceId = getStringInput("resourceId", "Enter the Filter Set's desired resource ID");
    String startDate = getStringInput("startDate", "Enter the start date in YYYYMMDD format",
        df.print(defaultStartTime));
    String endDate = getStringInput("endDate", "Enter the end date in YYYYMMDD format",
        df.print(defaultEndTime));
    // Optional fields.
    String creativeId = getOptionalStringInput("creativeId",
        "Enter the creativeId on which to filter");
    Long dealId = getOptionalLongInput("dealId", "Enter the dealId on which to filter.");
    String environment = getOptionalStringInput("environment",
        "Enter the environment on which to filter");
    String format = getOptionalStringInput("format", "Enter the Format on which to filter");
    List<String> platforms = getOptionalStringListInput("platforms",
        "Enter a comma separated list of Platforms on which to filter");
    List<Integer> sellerNetworkIds = getOptionalIntListInput("sellerNetworkIds",
        "Enter a comma separated list of integer seller network IDs on which to filter");
    String timeSeriesGranularity = getOptionalStringInput("timeSeriesGranularity",
        "Enter the TimeSeriesGranularity representing the desired granularity of time intervals "
            + "(if a time series breakdown is desired)");
    boolean isTransient = getBooleanInput("isTransient",
        "Enter a boolean representing whether the Filter Set is transient", true);

    String name = String.format("bidders/%s/accounts/%s/filterSets/%s", bidderResourceId,
        accountResourceId, resourceId);
    String ownerName = String.format("bidders/%s/accounts/%s", bidderResourceId, accountResourceId);

    FilterSet filterSet = new FilterSet();
    filterSet.setName(name);
    filterSet.setAbsoluteDateRange(buildAbsoluteDateRange(startDate, endDate));

    if(creativeId != null) {
      filterSet.setCreativeId(creativeId);
    }
    if(dealId != null) {
      filterSet.setDealId(dealId);
    }
    if(environment != null) {
      filterSet.setEnvironment(environment);
    }
    if(format != null) {
      filterSet.setFormat(format);
    }
    if(platforms != null) {
      filterSet.setPlatforms(platforms);
    }
    if(sellerNetworkIds != null) {
      filterSet.setSellerNetworkIds(sellerNetworkIds);
    }
    if(timeSeriesGranularity != null) {
      filterSet.setTimeSeriesGranularity(timeSeriesGranularity);
    }

    filterSet = adXClient.bidders().accounts().filterSets().create(ownerName, filterSet)
        .setIsTransient(isTransient).execute();

    System.out.println("========================================");
    System.out.printf("Filter Set created for Account with name \"%s\"%n", ownerName);
    System.out.println("========================================");
    System.out.printf("Filter Set name: %s%n", filterSet.getName());
    AbsoluteDateRange absDateRange = filterSet.getAbsoluteDateRange();
    System.out.println("AbsoluteDateRange");
    System.out.printf("\tStart date: %s%n",
        convertDateToString(absDateRange.getStartDate()));
    System.out.printf("\tEnd date: %s%n",
        convertDateToString(absDateRange.getEndDate()));
    timeSeriesGranularity = filterSet.getTimeSeriesGranularity();
    if(timeSeriesGranularity != null) {
      System.out.printf("Time series granularity: %s%n", timeSeriesGranularity);
    }
    creativeId = filterSet.getCreativeId();
    if(creativeId != null) {
      System.out.printf("Creative ID: %s%n", creativeId);
    }
    dealId = filterSet.getDealId();
    if(dealId != null) {
      System.out.printf("Deal ID: %s%n", dealId);
    }
    format = filterSet.getFormat();
    if(format != null) {
        System.out.printf("\tFormat: %s%n", format);
    }
    environment = filterSet.getEnvironment();
    if(environment != null) {
      System.out.printf("Environment: %s%n", environment);
    }
    platforms = filterSet.getPlatforms();
    if(platforms != null) {
      System.out.println("Platforms:");
      for(String platform : platforms) {
        System.out.printf("\t%s%n", platform);
      }
    }
    sellerNetworkIds = filterSet.getSellerNetworkIds();
    if(filterSet.getSellerNetworkIds() != null) {
      System.out.println("Seller network IDS:");
      for(Integer sellerNetworkId : sellerNetworkIds) {
        System.out.printf("\t%d%n", sellerNetworkId);
      }
    }
  }

  private AbsoluteDateRange buildAbsoluteDateRange(String startDate, String endDate) {
    DateTime s = df.parseDateTime(startDate);
    DateTime e = df.parseDateTime(endDate);

    if(e.isBefore(s.getMillis())) {
      throw new IllegalArgumentException("Error: The end date can not be before the start date.");
    }

    Date start = new Date();
    start.setYear(s.getYear());
    start.setMonth(s.getMonthOfYear());
    start.setDay(s.getDayOfMonth());
    Date end = new Date();
    end.setYear(e.getYear());
    end.setMonth(e.getMonthOfYear());
    end.setDay(e.getDayOfMonth());

    AbsoluteDateRange absoluteDateRange = new AbsoluteDateRange();
    absoluteDateRange.setStartDate(start);
    absoluteDateRange.setEndDate(end);

    return absoluteDateRange;
  }

  private String convertDateToString(Date date) {
    return String.format("%d%02d%02d", date.getYear(), date.getMonth(), date.getDay());
  }
}
